pkg/agent/manager/subscriber.go

```diff
@@ -3,30 +3,64 @@ package manager
 import (
 	"github.com/satori/go.uuid"
 	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/selector"
 )
 
 type subscriber struct {
-	c    chan []cache.Entry
+	c    chan *cache.Entry
 	sel  cache.Selectors
 	done chan struct{}
 }
 
-// subscribers is a map keyed by the Cache key, with a vaule mapped by Subscriber ID.
-type subscribers map[string]map[string]subscriber
+//type sID string
+// subscribers is a map keyed by the string representation of the selector sets, with a value mapped by Subscriber ID.
+// used to maintain a subscription of workloads
 
-func (s *subscribers) Add(sub subscriber) error {
+type subscribers struct {
+	selMap map[string][]uuid.UUID
+	sidMap map[uuid.UUID]*subscriber
+}
+
+func (s *subscribers) Add(sub *subscriber) error {
 
-	sID, err := uuid.NewV4()
+	id, err := uuid.NewV4()
 	if err != nil {
 		return err
 	}
+	s.sidMap[id] = sub
+
+	selSet := selector.NewSet(sub.sel)
+	selPSet := selSet.Power()
+	for sel := range selPSet {
+		selStr := sel.String()
+		s.selMap[selStr] = append(s.selMap[selStr], id)
+	}
 
-	// TODO: Remove the following line of code.
-	println(sID.String())
-	// s[sub.sel.Key] = append(s[sub.sek.Key], sub)
 	return nil
 }
 
-// func (*s subscribers) Remove(key string, id string) {
-// 	delete(s[key])
-// }
+func (s *subscribers) Notify(entry *cache.Entry) {
```

Comment:
Similar comment here about workload api having to return the full set of information. Here's the relavent snippet, but I'd still recommend you have a read over the spec - it's not as long as it looks :)

`
Every response message sent by the server MUST include the full set of information, and not just the information which has changed. This avoids complexity associated with state tracking on both Client and Server implementations, including the need for anti-entropy mechanisms.
`